static/app/components/forms/compactSelect.tsx

```diff
@@ -157,7 +164,23 @@ function CompactSelect({
   // Update the button label when the value changes
   function getLabel(newValue): React.ReactNode {
     const valueSet = Array.isArray(newValue) ? newValue : [newValue];
-    const optionSet = valueSet.map(val => options.find(opt => opt.value === val));
+
+    // Recursively finds the selected option(s) from the options list
+    function getSelectedOptions(
+      opts: Props<OptionType>['options'],
+      value: Props<OptionType>['value']
+    ): Props<OptionType>['options'] {
+      return opts.reduce((acc: Props<OptionType>['options'], cur) => {
+        if (cur.options) {
+          return [...acc, ...getSelectedOptions(cur.options, value)];
```

Comment:
Nit: There is no need to spread the acc on each iteration. Because you do that, this becomes an On^2 operation as each element gets touched once per loop. You can instead acc.concat(getSelectedOptions(cur.options, value)) and keep this an On.  Worth noting that it probably doesnt make a difference on small array sizes